design/ceph-openshift-fencing-mitigation.md

```diff
@@ -0,0 +1,72 @@
+
+Once [MachineHealthCheck controller](https://github.com/openshift/machine-api-operator#machine-healthcheck-controller) detects that a node is `NotReady` (or some other configured condition), it will remove the associated `Machine` which will cause the node to be deleted. The `MachineSet` controller will then replace the `Machine` via the machine-api. The exception is on baremetal platforms where fencing will reboot the underlying `BareMetalHost` object instead of deleting the `Machine`.
+
+
+## Why can't we use `PodDisruptionBudget`?
+
+Fencing does not use the eviction api. It is for `Machine`s and not `Pod`s.
+
+## Will we need to do large storage rebalancing after fencing?
+
+Hopefully not. On cloud platforms, the OSDs can be rescheduled on new nodes along with their backing PVs, and on baremetal where the local PVs are tied to a node, fencing will simply reboot the node instead of destroying it.
+
+
+# Problem Statement
+We need to ensure that only one node can be fenced at a time and that Ceph is fully recovered (has PGs clean) before any fencing is initiated. The available pattern for limiting fencing is the MachineDisruptionBudget which allows us to specify maxUnavailable. However, this won’t be sufficient to ensure that Ceph has recovered before fencing is initiated as MachineHealthCheck does not check anything other than the node state.
+
+Therefore, we will control how many nodes match the MDB by dynamically adding and removing labels as well as dynamically updating the MDB. By manipulating the MDB into a state where desiredHealthy > currentHealthy, we can disable fencing on the nodes the MDB points to.
+
+# Design:
+
+We will implement two controllers `machinedisruptionbudget-controller` and the `machine-controller` to be implemented through the controller pattern describere [here](https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg#hdr-Controller_Writing_Tips). Each controller watches a set of object kinds and reconciles one.
+
+The bottom line is that fencing is blocked if the PG state in not active+clean, but fencing continues on `Machine`s without the label which indicates that OSD resources are running there.
+
+## machinedisruptionbudget-controller
+This controller watches ceph PGs and CephClusters. We will ensure the reconciler is enqueued every 60s. It ensures that each CephCluster has a MDB created, and the MDB's value of maxUnvailable reflects the health of the Ceph Cluster's PGs.
+If all PGs are clean, maxUnavailable = 1.
+else, maxUnavailable = 0.
+
+We can share a ceph health cache with the other controller-runtime reconcilers that have to watch the PG "cleanliness".
+
+The MDB will target `Machine`s selected by a label maintainted by the `machine-controller`. The label is `fencegroup.rook.io/<cluster-name>`. 
```

Comment:
How about `fencegroup.rook.io/cluster: <cluster-name>` @rohantmp ?

Reply:
That works!

Reply:
I am wondering can it cause some issues when rook will be orchestrating multiple clusters.

Reply:
Do you mean multiple clusters sharing the same nodes? Yes, I'd also like to get the behavior better defined for the case where a Machine is selected by multiple MDBs.

For now, I think that we can just document the ambiguity. For MDBs managed by multiple entities to work, I think a basic requirement is that a machine can be referred to by more than one MDB. Unlike Pods and PDBs where only one application owns a Pod, a Machine/Node pair have multiple applications sharing it.

Reply:
Are PDBs and MDBs cluster-wide resources or do they belong to a namespace? If they're in a namespace, we are fine. If they are cluster-wide, we hopefully just need to make sure the PDB and MDB names are unique so clusters don't collide.

Reply:
PDBs are namespaced and MDBs are cluster-wide. We can prefix the name with the cluster-namespace.